Dockerfile

```diff
@@ -1,34 +1,36 @@
 # Simple usage with a mounted data directory:
 # > docker build -t gaia .
-# > docker run -v $HOME/.gaiad:/root/.gaiad gaia init
-# > docker run -v $HOME/.gaiad:/root/.gaiad gaia start
-
-FROM alpine:edge
+# > docker run -it -p 46657:46657 -p 46656:46656 -v ~/.gaiad:/root/.gaiad -v ~/.gaiacli:/root/.gaiacli gaia gaiad init
+# > docker run -it -p 46657:46657 -p 46656:46656 -v ~/.gaiad:/root/.gaiad -v ~/.gaiacli:/root/.gaiacli gaia gaiad start
+FROM golang:alpine AS build-env
 
 # Set up dependencies
-ENV PACKAGES go glide make git libc-dev bash
-
-# Set up GOPATH & PATH
-
-ENV GOPATH       /root/go
-ENV BASE_PATH    $GOPATH/src/github.com/cosmos
-ENV REPO_PATH    $BASE_PATH/cosmos-sdk
-ENV WORKDIR      /cosmos/
-ENV PATH         $GOPATH/bin:$PATH
-
-# Link expected Go repo path
+ENV PACKAGES make git libc-dev bash
 
-RUN mkdir -p $WORKDIR $GOPATH/pkg $ $GOPATH/bin $BASE_PATH
+# Set working directory for the build
+WORKDIR /go/src/github.com/cosmos/cosmos-sdk
 
 # Add source files
-
-ADD . $REPO_PATH
+COPY . .
 
 # Install minimum necessary dependencies, build Cosmos SDK, remove packages
 RUN apk add --no-cache $PACKAGES && \
-    cd $REPO_PATH && make get_tools && make get_vendor_deps && make build && make install && \
+    make get_tools && \
+    make get_vendor_deps && \
+    make build && \
+    make install && \
     apk del $PACKAGES
 
-# Set entrypoint
+# Final image
+FROM alpine:edge
+
+# Install ca-certificates
+RUN apk add --update ca-certificates
+WORKDIR /root
+
+# Copy over binaries from the build-env
+COPY --from=build-env /go/bin/gaiad /usr/bin/gaiad
+COPY --from=build-env /go/bin/gaiacli /usr/bin/gaiacli
```

Comment:
This makes more sense, but we have to make sure that gaiad is statically built or else it's not going to work under alpine.

Reply:
I've tested this and haven't run into any problems yet. I might not have hit the C code tho...

Reply:
Did you check if gaiad was statically built? `file gaiad`
If it is dynamically built, then the binary will only work with the same type of C library as where it was built. This will cause problems if something else (or an incompatible version) is installed on the target system. This is regardless of using C code or not, because the libraries need to be available during runtime.

For example, if you build on Ubuntu (glibc) and package it into an alpine image (muslc), your container will not run.

The tendermint binary was forced to be built statically, to avoid these issues. (It also helps with deterministic running too.)